File: kotlin-node/karakum/src/jsMain/resources/plugins/convertNodeJsQualifiedName.js
import ts from "typescript";

export default function (node, context, render) {
    if (
        ts.isQualifiedName(node)
        && ts.isIdentifier(node.left)
        && node.left.text === "NodeJS"
    ) {
        if (node.right.text === "Iterator") {
            return `js.iterable.JsIterator`
        }
        if (node.right.text === "AsyncIterator") {
            return `js.iterable.AsyncIterator`
        }
        if (node.right.text === "ArrayBufferView") {
            return `js.buffer.ArrayBufferView`
        }
        if (node.right.text === "TypedArray") {
            return `js.typedarrays.TypedArray`
        }
        if (node.right.text === "EventEmitter") {
            return `node.events.EventEmitter`
        }
        if (node.right.text === "Module") {
            return `node.module.Module`
        }
        if (node.right.text === "Require") {
            return `node.module.Require`
        }
        if (node.right.text === "Signals") {
            return `node.process.Signals`
        }
        if (node.right.text === "ProcessEnv") {
            return `node.process.ProcessEnv`
        }
        if (node.right.text === "Platform") {
            return `node.process.Platform`
        }

        return `node.${render(node.right)}`
    }

    if (
        ts.isPropertyAccessExpression(node)
        && ts.isIdentifier(node.expression)
        && node.expression.text === "NodeJS"
    ) {
        return `node.${render(node.name)}`
    }

    return null
}

File: kotlin-node/karakum/src/jsMain/kotlin/node/karakum/plugins/convertNodeJsQualifiedName.kt
package node.karakum.plugins

import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.createPlugin
import typescript.isIdentifier
import typescript.isPropertyAccessExpression
import typescript.isQualifiedName

val convertNodeJsQualifiedName = createPlugin { node, _, render ->
    nullable {
        ensure(isQualifiedName(node))

        val left = node.left
        ensure(isIdentifier(left))
        ensure(left.text == "NodeJS")

        when (node.right.text) {
            "Iterator" -> "js.iterable.JsIterator"
            "AsyncIterator" -> "js.iterable.AsyncIterator"
            "ArrayBufferView" -> "js.buffer.ArrayBufferView"
            "TypedArray" -> "js.typedarrays.TypedArray"

            "EventEmitter" -> "node.events.EventEmitter"

            "Module" -> "node.module.Module"
            "Require" -> "node.module.Require"

            "Process" -> "node.process.Process"
            "Architecture" -> "node.process.Architecture"
            "CpuUsage" -> "node.process.CpuUsage"
            "Platform" -> "node.process.Platform"
            "ProcessEnv" -> "node.process.ProcessEnv"
            "Signals" -> "node.process.Signals"
            else -> "node.${render(node.right)}"
        }
    } ?: nullable {
        ensure(isPropertyAccessExpression(node))

        val expression = node.expression
        ensure(isIdentifier(expression))
        ensure(expression.text == "NodeJS")

        val name = node.name
        ensure(isIdentifier(name))

        when (name.text) {
            "ProcessEnv" -> "node.process.ProcessEnv"
            else -> "node.${render(node.name)}"
        }
    }
}
